raster-foundry/raster-foundry
# [PAUSED] Lab frontend changes for Workspaces

## Overview
Frontend for lab workspaces

### Checklist

~- [ ] Styleguide updated, if necessary~
~- [ ] [Swagger specification](https://github.com/raster-foundry/raster-foundry-api-spec) updated~
~- [ ] Symlinks from new migrations present or corrected for any new migrations~

### Notes
This PR is being made before everything is finished in order to make organizing and tracking the work a bit easier.

### Issues to be pointed and carded
* Handle logic around creating, splitting. We should definite analyses by their root nodes and allow nodes to belong to multiple analyses.
* Linking and deleting nodes / links doesn't work correctly in some cases, especially nodes which are used as inputs to more than one other node
* ~~Stats don't work~~
* Histogram doesn't work
* When adding multiple nodes (template, MAML equation), nodes that have multiple outputs are sometimes placed incorrectly by the layout. This may be related to the first issue
* Zoom to fit is really funky - doesn't work on large stuff, negative coordinates cause it to go haywire, large coordinates can also cause issues. I don't think this was introduced by my changes, but it's a lot more obvious now that we can actively add stuff to the diagram
* Allow previewing analyses which have all inputs filled, independent of other analyses
* Occasional syncing issue between server and client node positions

## Testing Instructions
 *  Create a workspace
* Create a template
* Create a workspace from the template detail page, and verify that the workspace is generated with the chosed template
* Add a template to a workspace from the sidebar
* Add a MAML equation to a workspace from the sidebar
* Add an input and function node to the workspace from the sidebar
* Link a function to two inputs (it's a bit finnicky due to the bad logic around it at the moment, and the link won't show up until you refresh the page)
* On a workspace with two analyses (import two templates), verify that previewing and comparing across analyses works as expected
* Nodes cannot be previewed until all inputs are vaild
* Nodes and workspaces can be renamed

Works towards #3032 
Closes #3035 
Closes #3044 
Closes #3033 
Closes #3165
